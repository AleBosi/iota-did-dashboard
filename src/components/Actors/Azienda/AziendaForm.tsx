import React, { useState } from "react";
import { Azienda } from "../../../models/azienda";

interface Props {
  onCreate: (azienda: Azienda) => void;
}

export default function AziendaForm({ onCreate }: Props) {
  const [name, setName] = useState("");
  const [vat, setVat] = useState("");
  const [lei, setLei] = useState("");
  const [address, setAddress] = useState("");
  const [email, setEmail] = useState("");
  const [country, setCountry] = useState("");

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const azienda: Azienda = {
      id: "did:iota:evm:" + Date.now(), // replace with generateDid()
      name,
      seed: "SEED_" + Date.now(),       // replace with generateSeed()
      legalInfo: {
        vat,
        lei,
        address,
        email,
        country
      },
      creators: [],
      operatori: [],
      macchinari: [],
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    };
    onCreate(azienda);
    setName(""); setVat(""); setLei(""); setAddress(""); setEmail(""); setCountry("");
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-col gap-2">
      <input placeholder="Ragione sociale" value={name} onChange={e => setName(e.target.value)} required />
      <input placeholder="Partita IVA" value={vat} onChange={e => setVat(e.target.value)} />
      <input placeholder="LEI" value={lei} onChange={e => setLei(e.target.value)} />
      <input placeholder="Indirizzo" value={address} onChange={e => setAddress(e.target.value)} />
      <input placeholder="Email" value={email} onChange={e => setEmail(e.target.value)} />
      <input placeholder="Nazione" value={country} onChange={e => setCountry(e.target.value)} />
      <button type="submit" className="bg-blue-500 text-white px-4 py-1 rounded">Crea azienda</button>
    </form>
  );
}
