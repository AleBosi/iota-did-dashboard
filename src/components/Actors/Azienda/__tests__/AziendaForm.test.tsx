import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import AziendaForm from "../AziendaForm";

describe("AziendaForm", () => {
  const mockOnCreate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renderizza tutti i campi richiesti", () => {
    render(<AziendaForm onCreate={mockOnCreate} />);
    expect(screen.getByPlaceholderText(/ragione sociale/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/partita iva/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/lei/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/indirizzo/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/email/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/nazione/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /crea azienda/i })).toBeInTheDocument();
  });

  test("non chiama onCreate se il campo Ragione sociale Ã¨ vuoto", async () => {
    render(<AziendaForm onCreate={mockOnCreate} />);
    fireEvent.click(screen.getByRole("button", { name: /crea azienda/i }));
    await waitFor(() => {
      expect(mockOnCreate).not.toHaveBeenCalled();
    });
  });

  test("chiama onCreate con i dati inseriti", async () => {
    render(<AziendaForm onCreate={mockOnCreate} />);
    fireEvent.change(screen.getByPlaceholderText(/ragione sociale/i), { target: { value: "Azienda Test" } });
    fireEvent.change(screen.getByPlaceholderText(/partita iva/i), { target: { value: "12345678901" } });
    fireEvent.change(screen.getByPlaceholderText(/lei/i), { target: { value: "LEI123456" } });
    fireEvent.change(screen.getByPlaceholderText(/indirizzo/i), { target: { value: "Via Roma, 1" } });
    fireEvent.change(screen.getByPlaceholderText(/email/i), { target: { value: "info@test.it" } });
    fireEvent.change(screen.getByPlaceholderText(/nazione/i), { target: { value: "Italia" } });

    fireEvent.click(screen.getByRole("button", { name: /crea azienda/i }));

    await waitFor(() => {
      expect(mockOnCreate).toHaveBeenCalledTimes(1);
      const azienda = mockOnCreate.mock.calls[0][0];
      expect(azienda.name).toBe("Azienda Test");
      expect(azienda.legalInfo.vat).toBe("12345678901");
      expect(azienda.legalInfo.lei).toBe("LEI123456");
      expect(azienda.legalInfo.address).toBe("Via Roma, 1");
      expect(azienda.legalInfo.email).toBe("info@test.it");
      expect(azienda.legalInfo.country).toBe("Italia");
      expect(azienda.id).toMatch(/^did:iota:evm:/);
      expect(azienda.did).toMatch(/^did:iota:evm:/);
      expect(azienda.seed).toMatch(/^SEED_/);
      expect(Array.isArray(azienda.creators)).toBe(true);
      expect(Array.isArray(azienda.vcIds)).toBe(true);
          });
  });

  test("svuota i campi dopo il submit", async () => {
    render(<AziendaForm onCreate={mockOnCreate} />);
    fireEvent.change(screen.getByPlaceholderText(/ragione sociale/i), { target: { value: "Azienda Test" } });
    fireEvent.click(screen.getByRole("button", { name: /crea azienda/i }));

    await waitFor(() => {
      expect(screen.getByPlaceholderText(/ragione sociale/i)).toHaveValue("");
    });
  });

  // Snapshot di base
  test("matcha lo snapshot", () => {
    const { asFragment } = render(<AziendaForm onCreate={mockOnCreate} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
