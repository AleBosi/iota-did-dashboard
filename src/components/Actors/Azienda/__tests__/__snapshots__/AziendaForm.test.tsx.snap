// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AziendaForm matcha lo snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-2"
  >
    <input
      placeholder="Ragione sociale"
      required=""
      value=""
    />
    <input
      placeholder="Partita IVA"
      value=""
    />
    <input
      placeholder="LEI"
      value=""
    />
    <input
      placeholder="Indirizzo"
      value=""
    />
    <input
      placeholder="Email"
      value=""
    />
    <input
      placeholder="Nazione"
      value=""
    />
    <button
      class="bg-blue-500 text-white px-4 py-1 rounded"
      type="submit"
    >
      Crea azienda
    </button>
  </form>
</DocumentFragment>
`;
